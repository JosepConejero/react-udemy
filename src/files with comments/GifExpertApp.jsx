import { useState } from "react";
import { AddCategory } from "./componentes/AddCategory";
import { GifGrid } from "./componentes/GifGrid";

export const GifExpertApp = () => {
  const [categories, setCategories] = useState(["One Punch"]);

  /*  const onAddCategory = () => {
    //setCategories([...categories, "Valorant"]);
    setCategories((cat) => [...cat, "Valorant"]);
  }; */

  const onAddCategory = (newCategory) => {
    //setCategories((cat) => [...cat, newCategory]);
    if (categories.includes(newCategory)) return; // si encuentra la categoría, no la añade otra vez
    setCategories([newCategory, ...categories]);
  };

  return (
    <>
      {/*título*/}
      <h1>GifExpertApp</h1>
      {/*input*/}
      {/* <AddCategory categories={categories} setCategories={setCategories} /> */}
      <AddCategory onNewCategory={(value) => onAddCategory(value)} />
      {/* mando la referencia a la función*/}
      {/*listado de gif*/}
      {/*  <button onClick={onAddCategory}>Agregar</button> */}
      {/* <ol> */}
      {categories.map((category) => (
        /* {<li key={category}>{category}</li>} */
        /*   <div key={category}>
            <h3>{category}</h3>
            <li>{category}</li>
          </div> */
        <GifGrid key={category} category={category} />
      ))}
      {/* </ol> */}
      {/*git item*/}
    </>
  );
};
