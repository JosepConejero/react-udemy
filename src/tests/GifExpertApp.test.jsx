/* eslint-disable testing-library/no-debugging-utils */
import { fireEvent, render, screen } from "@testing-library/react";
import { GifExpertApp } from "../GifExpertApp";
//import { useState } from "react";

//jest.mock("react");

describe("Pruebas en el componente <GifExpertApp/>", () => {
  test("Deberá hacer match con el snapshot", () => {
    console.log(
      "==============================================================="
    );
    const { container } = render(<GifExpertApp />);
    expect(container).toMatchSnapshot();
  });

  test("Deberá tener el encabezado GifExpertApp", () => {
    render(<GifExpertApp />);
    expect(screen.getAllByRole("heading", { level: 1 }).length).toBe(1);
    expect(screen.getByText("GifExpertApp")).toBeTruthy();
  });

  test("Deberá añadir una categoría", () => {
    const newCategory = "One Punch";
    const initialCategory = "dragon ball";

    render(<GifExpertApp />);
    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.input(input, { target: { value: newCategory } });
    fireEvent.submit(form);

    expect(screen.getAllByText(/dragon ball/i).length).toBe(1);
    expect(screen.getByText(initialCategory).innerHTML).toBe("dragon ball");
    expect(screen.getByText(initialCategory)).not.toBeNull();
    // expect(screen.getByText("dragon ball")).toBeInTheDocument();
    expect(screen.getAllByRole("heading", { level: 3 }).length).toBe(2);

    expect(screen.getAllByRole("heading", { level: 3 })[0].innerHTML).toBe(
      newCategory
    );
    expect(screen.getAllByRole("heading", { level: 3 })[1].innerHTML).toBe(
      initialCategory
    );
  });

  test("No deberá añadir una categoría si ya existe", () => {
    render(<GifExpertApp />);
    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.input(input, { target: { value: "One Punch" } });
    fireEvent.submit(form);

    fireEvent.input(input, { target: { value: "One Punch" } });
    fireEvent.submit(form);

    expect(screen.getAllByRole("heading", { level: 3 }).length).toBe(2);
  });

  test("No deberá añadir una categoría si está vacía", () => {
    render(<GifExpertApp />);
    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.input(input, { target: { value: "" } });
    fireEvent.submit(form);
    //screen.debug();

    expect(screen.getAllByRole("heading", { level: 3 }).length).toBe(1);
  });
});
