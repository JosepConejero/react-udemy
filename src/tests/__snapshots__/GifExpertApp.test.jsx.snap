// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifExpertApp/> Deber√° hacer match con el snapshot 1`] = `
<div>
  <h1>
    GifExpertApp
  </h1>
  <form
    aria-label="formElement"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
  <h3>
    dragon ball
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
